### DISCUSSION TITLE
Allow implicit usage of static interface members

### DISCUSSION THREAD

**Discussion by koszeggy** [dotnet/csharplang#5265] [Language Ideas]
**Disclaimer:** I couldn't decide whether it should be a compiler bug report, a language proposal or an API proposal so in the end I created this as a discussion. Feel free to migrate it to the appropriate repo as a desired type of issue.

### The Issue

With preview features [enabled](https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/) consider the following code that tries to utilize the power of static interface members (#4436):

```cs
Half h1 = (Half)1f;
Half h2 = (Half)2f;
Half h3;

// CS0019: Operator '+' cannot be applied to operands of type 'Half' and 'Half'
h3 = h1 + h2;
```

The code above produces CS0019 because [Half](https://docs.microsoft.com/en-us/dotnet/api/system.half) implements the `+` operator only via its `IAdditionOperators` interface [implementation](https://github.com/dotnet/runtime/blob/3de620787ab3962dcb89664a691cf917653d5bd2/src/libraries/System.Private.CoreLib/src/System/Half.cs#L710) and not as a "regular" operator.

#### Workaround attempt:
```cs
// Alert: boxing occurs
INumber<Half> n1 = h1;
INumber<Half> n2 = h2;

// CS0019: Operator '+' cannot be applied to operands of type 'INumber<Half>' and 'INumber<Half>'
// candidates are:
//   System.Half +(System.Half, System.Half, System.Half) in interface 'IAdditionOperators<Half,Half,Half>'
h3 = n1 + n2;
```

The error and the suggestion about the candidates look particularly silly together... it almost seems as if this should be allowed. Though I know that this would allow accessing static members via the interface, which is not quite desired (not mentioning the boxing).

#### Working workaround:
```cs
// we have to create a generic method (can be a local function, too):
static T AddByGenericMethod<T>(T n1, T n2) where T : IAdditionOperators<T,T,T> => n1 + n2;

h3 = AddByGenericMethod(h1, h2);
```

### Proposal A (the preferred one)

Simply allow  `h1 + h2` if the operands implement `IAdditionOperators`. The `Half` struct does it by [implementing](https://github.com/dotnet/runtime/blob/3de620787ab3962dcb89664a691cf917653d5bd2/src/libraries/System.Private.CoreLib/src/System/Half.cs#L23) the `IBinaryFloatingPoint<Half>` interface.

And of course, this should apply all of the other operators, too.

### Proposal B

Allow some way (eg. by explicit casting) to invoke the operators, without the need of writing new generic methods (like in the failed attempt above).

### Proposal C (API)

Provide an API with all those predefined generic methods so we don't need to write them again and again:

```cs
h3 = Math<Half>.Add(h1, h2);
```

***Note:*** Instead of a new `Math<TNumber>` class we would have new generic overloads in the good old `Math` class but as we already have also a `MathF` class, I just didn't feel it quite correct.

### Proposal D (the least preferred one)

Add the non-interface operator implementations to each and every types one by one. An analyzer warning should be triggered if a type implements operators via interfaces but does not have regular operator implementations.

----

### Edit:

Non-operator static members have the same issue. For example, it should be allowed:

```cs
// CS8926: A static abstract interface member can be accessed only on a type parameter.
Half h4 = INumber<Half>.Abs(h3);
```

Why is the compiler so picky? Even though `INumber<T>.Abs` is `abstract`, we specified `Half` so the call is not ambiguous. To satisfy the compiler we must use a generic method again so we will have a required type argument to invoke the `Abs` method on:

```cs
// Invoking INumber.Abs<Half> by a dummy generic method to be able to use 'T.Abs()' as demanded (can be a local function):
static T AbsByGenericMethod<T>(T value) where T : INumber<T> => T.Abs(value);
Half h4 = AbsByGenericMethod(h3);
```

**Comment by tannergooding** [comment]
There are a few things to address in the above:
* `Half` explicitly implements these interfaces only for .NET 6 where the feature is in preview. The plan is to have it implicitly implement these interfaces when the feature becomes stable and so they will be "just available".
  * This was done for several reasons, including to help prevent accidental dependence on an experimental feature and experimental API surface area
* Where a type explicitly implements an interface; it would not be a good change to simply allow the "hidden" interface members to just be accessed via the type. The author of the type made them non-public for a reason, whether the consumer thinks it good or bad.
  * This would likely also be a breaking change due to ambiguities and potential other issues that could come up
  * If the interface is implicitly implemented, then the members are already just available so nothing needs to be done

The current proposal on how to address this for explicitly implemented interface members (which is on me to do the proposal writeup for) is for the language to properly recognize the "self constraint". That is, the signature of something like `interface IBinaryNumber<TSelf>` will likely change to something similar to `interface IBinaryNumber<this TSelf>` (this is just an example syntax; there are several alternatives proposed and none have been decided on).

This special syntax would place an additional constraint on the language that `ISomeInterface<this T>` can only be implemented by `T`. That is, `struct MyStruct : ISomeInterface<int>` would be illegal, only `struct MyStruct : ISomeInterface<MyStruct>`  would be valid.

Once this is in place, the language can make assumptions about the given generic parameter and so `IBinaryNumber<int>.IsPow2`, for example, could be known to exactly resolve to the method exposed by `int`. This allows the language to validate that `int` does implement the interface (implicitly or explicitly) and still allow access to the static methods exposed by it via the interface; much as you would with any other interface method.

**Reply by koszeggy** [reply]
> Once this is in place, the language can make assumptions about the given generic parameter and so `IBinaryNumber<int>.IsPow2`, for example, could be known to exactly resolve to the method exposed by `int`.

Is this still planned? Because 5 years later, this still fails with CS8926.

While testing the new `Decimal##` types in .NET 11 I realized that unlike the new decimal types, the regular `Decimal` struct implements a lot of static interface members only explicitly (e.g. `Pi`), but `IFloatingPointConstants<decimal>.Pi` still cannot be used directly.

**Comment by GeeLaw** [comment]
> Proposal A "Simply allow `h1 + h2` if the operands implement `IAdditionOperators`."

This is undesirable because it makes `IAdditionOperators` special. A user could define

```csharp
interface I<T> where T : I<T>
{
  static abstract T operator +(T a, T b);
}
```

but if `h` is typed non-type-parameter `T` and `T` implements `I<T>` explicitly, C# wouldn't allow you to say `h + h`.

---

> Proposal D (the least preferred one) "An analyzer warning should be triggered if a type implements operators via interfaces but does not have regular operator implementations."

This is a very bad idea. C# makes it easy to implement interfaces implicitly, so the fact that a type has implemented an interface member explicitly has strong signal of design intention.

Please also note that an interface need not be as accessible as types implementing it, so an explicitly implemented member might be effectively `internal`. Allowing access without being able to name the interface (in the generics helper example, it's done by being able to state the interface as a constraint) would break this expectation.

---

> Why is the compiler so picky? Even though `INumber<T>.Abs` is abstract, we specified `Half` so the call is not ambiguous.

The static member belongs to a type. If you say `INumber<Half>.Abs`, it means the version of `Abs` on `INumber<Half>`, which is `abstract` hence cannot be invoked. You must say `Half.Abs` (if `Half` implements this member implicitly).

Also, consider legal although contrived code:

```csharp
interface I<T> where T : I<T> { static abstract T Foo { get; } }
struct A : I<A> { /* impl omitted */ }
struct B : I<A> { /* impl omitted */ }

static void Bar<T, U>() where T : I<U> where U : I<U>
{
  // T.Foo and U.Foo are different!
}
```

---

Re @tannergooding on `this` type parameter... (https://github.com/dotnet/csharplang/discussions/5265#discussioncomment-1455474) I'm not sure about the latest idea on the exact semantics, but I would presume `I<this T>` means `T : I<T>` and `obj is I<T>` implies `obj is T` (and no more than these two conditions), which does not prevent `static abstract I<T>.M` from having two implementations, because `T` can be a non-sealed type and derived types can either override the implementation in `T` (if `T` makes it `virtual`) or re-implement `I<T>`.

**Reply by tannergooding** [reply]
APIs that are marked explicit are intentionally hidden on the type. If a user wants to access them, they must go through the interface. This is nothing new and is incredibly trivial to achieve for the cases that do need it.

For the .NET Libraries, most explicitly implemented members are "constant" in nature (always throwing, always returning a specific value, being a no-op, etc) or have a more suitable public alternative. If there are outliers, then feel free to open an API proposal asking it be publicly exposed instead.

**Reply by koszeggy** [reply]
@GeeLaw Sorry if I confused you by necromancing a 5 years old thread (my very 1st one, actually 😊). `h1 + h2` works now; not because considering `IAdditionOperators`, but since `Half` received the regular operators since then. Hence the original proposals are out of question now, but thank you for taking the opportunity to tell why they would have been a very bad idea anyway.

My follow-up question regarded only what I quoted. I did so, because @tannergooding predicted that something like `IBinaryNumber<int>.IsPow2` would be supported later on, which is similar to my recent finding with `IFloatingPointConstants<decimal>.Pi`, which now requires a helper method to access it (unlike for the new decimal types where these are publicly implemented). So my question is really what it is: "is this still planned?", without even proposing/demanding to make it happen. I know that the [generic bridge](https://github.com/dotnet/csharplang/discussions/6308) feature is still in progress, so maybe just not everything is in place yet.
